cvat/apps/engine/utils.py

```diff
@@ -165,25 +165,22 @@ def define_dependent_job(
     if not should_be_dependent:
         return None
 
-    started_user_jobs = [
-        job
-        for job in queue.job_class.fetch_many(
-            queue.started_job_registry.get_job_ids(), queue.connection
-        )
-        if job and job.meta.get("user", {}).get("id") == user_id
-    ]
-    deferred_user_jobs = [
-        job
-        for job in queue.job_class.fetch_many(
-            queue.deferred_job_registry.get_job_ids(), queue.connection
-        )
+    keys = []
+    with queue.connection.pipeline() as pipeline:
         # Since there is no cleanup implementation in DeferredJobRegistry,
         # this registry can contain "outdated" jobs that weren't deleted from it
         # but were added to another registry. Probably such situations can occur
         # if there are active or deferred jobs when restarting the worker container.
-        if job and job.meta.get("user", {}).get("id") == user_id and job.is_deferred
+        for registry in [queue.started_job_registry, queue.deferred_job_registry]:
+            pipeline.zrange(registry.key, 0, -1)
+        pipeline.lrange(queue.key, 0, -1)
+        results = pipeline.execute()
+        keys = [bytes_key.decode('utf-8') for result in results for bytes_key in result]
```

Comment:
Why did you change this to use redis API instead of rq? Looks like we can receive expired jobs here since the`StartedJobRegistry.cleanup` method is not being called.

Reply:
- The first problem I found here was that queued jobs are not considered in this logic (what looks like error, as it allows a user to push as many jobs to **queued** (ignoring deferred) as he wants), if all workers are busy with other tasks of other users.

The initial fix was like:

```
started_user_jobs = [
        job
        for job in queue.job_class.fetch_many(
            queue.started_job_registry.get_job_ids(), queue.connection
        )
        if job and job.meta.get("user", {}).get("id") == user_id
    ]
deferred_user_jobs = [
    job
    for job in queue.job_class.fetch_many(
        queue.deferred_job_registry.get_job_ids(), queue.connection
    )
    # Since there is no cleanup implementation in DeferredJobRegistry,
    # this registry can contain "outdated" jobs that weren't deleted from it
    # but were added to another registry. Probably such situations can occur
    # if there are active or deferred jobs when restarting the worker container.
    if job and job.meta.get("user", {}).get("id") == user_id and job.is_deferred
]
queued_user_jobs = [
    job
    for job in queue.job_class.fetch_many(
        queue.get_job_ids(), queue.connection
    )
    if job and job.meta.get("user", {}).get("id") == user_id and job.is_deferred
]
all_user_jobs = started_user_jobs + deferred_user_jobs + queued_user_jobs
```

Reply:
Yes, because I was trying to reproduce it specially for you :)

Reply:
And checked one more time, with the current implementation `export:project.id128-by-admin` after updating the resource goes immediately to `queued` as result of logic with avoiding circular dependencies. 

That is also not good, but that is about using the same IDs.